internal/protoc/compiler.go

```diff
@@ -161,8 +161,40 @@ func (c *compiler) Compile(protoSet *file.ProtoSet) (*CompileResult, error) {
 			fileDescriptorSets = append(fileDescriptorSets, fileDescriptorSet)
 		}
 	}
+	// Given that we only allow one build configuration, we can merge all of
+	// file descriptor sets into a single set.
+	fds, err := mergeFileDescriptorSets(fileDescriptorSets)
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge file descriptors: %v", err)
+	}
 	return &CompileResult{
-		FileDescriptorSets: fileDescriptorSets,
+		FileDescriptorSet: fds,
+	}, nil
+}
+
+// Merge the file descriptor sets into a single set. All of the import paths should
+// refer to the same proto files, so we verify by comparing the raw bytes
+// between the files that are referenced more than once.
+func mergeFileDescriptorSets(sets []*descriptor.FileDescriptorSet) (*descriptor.FileDescriptorSet, error) {
+	// Used to determine whether or not a *descriptor.FileDescriptorProto
+	// has already been added to the merged set. We use the file descriptor's
+	// Name attribute to uniquely represent every file descriptor.
+	files := map[string]*descriptor.FileDescriptorProto{}
+	for _, s := range sets {
+		for _, f := range s.File {
+			name := f.GetName()
```

Comment:
I’m not confident this is 100% safe, I need to think about it more. The safest would be to marshal each FileDescriptorProto and hash the byte slice (cast to a string), and check if there are duplicates

Reply:
Yea, that's actually what we're doing here. `GetName` uniquely identifies the file and we use the `String` method to marshal that `FileDescriptorProto` into bytes and compare them. For more, see the description above.

Reply:
Oh whoops, missed the string thing! So just for clarity, it isn't technically marshalling to Protobuf, it's doing `proto.CompactTextString()` on the value. Should be the same effect in theory, just different serialization format, but for safety, I would worry that `CompactTextString` does not cover everything (with absolutely no research on the subject right now), and I would have preffered to do `proto.Marshal()` - just for future reference!